### PR TITLE
Add per-retry request logging hook

### PR DESCRIPTION
**Pull request by jen20**

### CODE REVIEW

File: client_test.go [hunk 1]
```diff
@@ -188,6 +200,38 @@ func TestClient_Get(t *testing.T) {
 	resp.Body.Close()
 }
 
+func TestClient_RequestLogHook(t *testing.T) {
+	// Mock server which always responds 500.
```
Comment (ryanuber) [inline client_test.go]: But does it? :) Looks like this comment is just outdated.

File: client_test.go [hunk 2]
```diff
@@ -188,6 +200,38 @@ func TestClient_Get(t *testing.T) {
...
+		if r.RequestURI != "/foo/bar" {
+			t.Fatalf("bad uri: %s", r.RequestURI)
+		}
+		w.WriteHeader(200)
+	}))
+	defer ts.Close()
+
+	retries := -1
+
+	client := NewClient()
```
Comment (ryanuber) [inline client_test.go]: Can we set a logger on the new client, write to it in the callback and check its buffer? Or at least just ensure the same logger instance is passed through.

**Comment by ryanuber** [comment]
@jen20 just minor comments but LGTM!

**Comment by ryanuber** [comment]
We should probably also test the callback with a failing HTTP request to ensure we get a callback for each retry.

**Comment by jen20** [comment]
Thanks for the review @ryanuber - all issues have been addressed. Another instance of the out-of-date comment is also removed in this.

**Comment by ryanuber** [comment]
LGTM! Merge away :ship:

**Pull request merged**